File: extract_server/src/extract_server/extraction/matching.py
from __future__ import annotations

import json
import math
import os
from dataclasses import dataclass
from typing import Any, Literal

from extract_server.extraction.scoring import normalize_name

STOP_TOKENS = frozenset({
    "with", "and", "the", "a", "an", "of", "for", "in", "on", "w", "ea", "pkg", "pack",
    "value", "fresh", "food", "foods", "mart", "label", "superior", "bridge",
})

GENERIC_PRODUCT_TOKENS = frozenset({
    "rice", "ball", "balls", "pork", "soy", "sauce", "milk", "egg", "eggs", "tofu",
    "chips", "tomato", "tomatoes", "pasta", "frozen", "fresh", "boneless", "whole",
    "thyme", "herb", "herbs", "spice", "spices",
})

PRODUCTION_EMB_WEIGHT = 0.75
PRODUCTION_TOK_WEIGHT = 0.25
# Grocery name embeddings cluster tightly (~0.70–1.0). Stretch that band to [0, 1]
# before applying emb_weight so unrelated pairs no longer dominate the composite.
DEFAULT_EMBED_COSINE_FLOOR = 0.70

MatchPath = Literal["barcode", "exact_name", "composite"]


def match_top_n() -> int:
    return int(os.environ.get("GROCERY_MATCH_TOP_N", "5"))


def match_min_score() -> float:
    return float(os.environ.get("GROCERY_MATCH_MIN_SCORE", "0.5"))


def embed_cosine_floor() -> float:
    return float(os.environ.get("GROCERY_EMBED_COSINE_FLOOR", str(DEFAULT_EMBED_COSINE_FLOOR)))


@dataclass(frozen=True)
class MatchSighting:
    id: str
    photo_id: str
    product_name: str
    category: str = ""
    brand: str | None = None
    barcode: str | None = None


@dataclass(frozen=True)
class MatchScoreDetail:
    final_score: float
    path: MatchPath
    embedding_cosine: float | None
    embedding_score: float
    cosine_floor: float
    token_jaccard: float
    emb_weight: float
    tok_weight: float
    normalized_source: str
    normalized_target: str
    formula: str | None


def tokenize(name: str) -> list[str]:
    return [t for t in normalize_name(name).split() if t and t not in STOP_TOKENS]


def token_jaccard(a: str, b: str, *, ignore_generic: bool = False) -> float:
    ta = {t for t in tokenize(a) if not (ignore_generic and t in GENERIC_PRODUCT_TOKENS)}
    tb = {t for t in tokenize(b) if not (ignore_generic and t in GENERIC_PRODUCT_TOKENS)}
    if not ta or not tb:
        return 0.0
    return len(ta & tb) / len(ta | tb)


def cosine_similarity(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(x * x for x in b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)


def rescale_cosine(cosine: float, *, floor: float | None = None) -> float:
    """Map cosine from [floor, 1] onto [0, 1], clamped."""
    c_min = embed_cosine_floor() if floor is None else floor
    if c_min >= 1.0:
        return 1.0 if cosine >= 1.0 else 0.0
    t = (cosine - c_min) / (1.0 - c_min)
    if t <= 0.0:
        return 0.0
    if t >= 1.0:
        return 1.0
    return t


def embedding_score(
    vector_a: list[float] | None,
    vector_b: list[float] | None,
    *,
    floor: float | None = None,
) -> float:
    if vector_a is None or vector_b is None:
        return 0.0
    return rescale_cosine(cosine_similarity(vector_a, vector_b), floor=floor)


def _composite_formula(emb_weight: float, tok_weight: float, cosine_floor: float) -> str:
    return (
        f"{emb_weight} * emb + {tok_weight} * tok "
        f"(emb = clamp((cos - {cosine_floor}) / (1 - {cosine_floor}), 0, 1))"
    )


def score_pair_detail(
    a: MatchSighting,
    b: MatchSighting,
    *,
    vector_a: list[float] | None = None,
    vector_b: list[float] | None = None,
    emb_weight: float = PRODUCTION_EMB_WEIGHT,
    tok_weight: float = PRODUCTION_TOK_WEIGHT,
    cosine_floor: float | None = None,
) -> MatchScoreDetail:
    norm_a = normalize_name(a.product_name)
    norm_b = normalize_name(b.product_name)
    floor = embed_cosine_floor() if cosine_floor is None else cosine_floor
    raw_cosine = (
        cosine_similarity(vector_a, vector_b)
        if vector_a is not None and vector_b is not None
        else None
    )
    emb = (
        rescale_cosine(raw_cosine, floor=floor)
        if raw_cosine is not None
        else 0.0
    )
    tok = token_jaccard(a.product_name, b.product_name, ignore_generic=True)

    if a.barcode and b.barcode and a.barcode == b.barcode:
        return MatchScoreDetail(
            final_score=1.0,
            path="barcode",
            embedding_cosine=raw_cosine,
            embedding_score=emb,
            cosine_floor=floor,
            token_jaccard=tok,
            emb_weight=emb_weight,
            tok_weight=tok_weight,
            normalized_source=norm_a,
            normalized_target=norm_b,
            formula=None,
        )

    if norm_a == norm_b:
        return MatchScoreDetail(
            final_score=1.0,
            path="exact_name",
            embedding_cosine=raw_cosine,
            embedding_score=emb,
            cosine_floor=floor,
            token_jaccard=tok,
            emb_weight=emb_weight,
            tok_weight=tok_weight,
            normalized_source=norm_a,
            normalized_target=norm_b,
            formula=None,
        )

    final = emb_weight * emb + tok_weight * tok
    return MatchScoreDetail(
        final_score=final,
        path="composite",
        embedding_cosine=raw_cosine,
        embedding_score=emb,
        cosine_floor=floor,
        token_jaccard=tok,
        emb_weight=emb_weight,
        tok_weight=tok_weight,
        normalized_source=norm_a,
        normalized_target=norm_b,
        formula=_composite_formula(emb_weight, tok_weight, floor),
    )


def score_pair(
    a: MatchSighting,
    b: MatchSighting,
    *,
    vector_a: list[float] | None = None,
    vector_b: list[float] | None = None,
) -> float:
    return score_pair_detail(a, b, vector_a=vector_a, vector_b=vector_b).final_score


def rank_related(
    source: MatchSighting,
    catalog: list[MatchSighting],
    *,
    vectors: dict[str, list[float]],
    top_n: int | None = None,
    min_score: float | None = None,
    emb_weight: float = PRODUCTION_EMB_WEIGHT,
    tok_weight: float = PRODUCTION_TOK_WEIGHT,
    exclude_same_photo: bool = False,
) -> list[tuple[str, float]]:
    details = rank_related_detail(
        source,
        catalog,
        vectors=vectors,
        top_n=top_n,
        min_score=min_score,
        emb_weight=emb_weight,
        tok_weight=tok_weight,
        exclude_same_photo=exclude_same_photo,
    )
    return [(item_id, detail.final_score) for item_id, detail in details]


def rank_related_detail(
    source: MatchSighting,
    catalog: list[MatchSighting],
    *,
    vectors: dict[str, list[float]],
    top_n: int | None = None,
    min_score: float | None = None,
    emb_weight: float = PRODUCTION_EMB_WEIGHT,
    tok_weight: float = PRODUCTION_TOK_WEIGHT,
    exclude_same_photo: bool = False,
) -> list[tuple[str, MatchScoreDetail]]:
    limit = top_n if top_n is not None else match_top_n()
    threshold = min_score if min_score is not None else match_min_score()
    scored: list[tuple[str, MatchScoreDetail]] = []
    source_vector = vectors.get(source.id)

    for candidate in catalog:
        if candidate.id == source.id:
            continue
        if exclude_same_photo and candidate.photo_id == source.photo_id:
            continue
        detail = score_pair_detail(
            source,
            candidate,
            vector_a=source_vector,
            vector_b=vectors.get(candidate.id),
            emb_weight=emb_weight,
            tok_weight=tok_weight,
        )
        if detail.final_score >= threshold:
            scored.append((candidate.id, detail))

    scored.sort(key=lambda item: (-item[1].final_score, item[0]))
    return scored[:limit]


def sighting_from_row(row: dict[str, Any]) -> MatchSighting:
    other = row.get("other") or {}
    if isinstance(other, str):
        other = json.loads(other or "{}")
    return MatchSighting(
        id=row["id"],
        photo_id=row["photo_id"],
        product_name=row["product_name"],
        category=str(other.get("category") or row.get("category") or ""),
        brand=other.get("brand") if isinstance(other.get("brand"), str) else None,
        barcode=other.get("barcode") if isinstance(other.get("barcode"), str) else None,
    )

File: experiments/match-explorer/src/lib/matching.ts
import type { MatchDetail, MatchSighting, MatchStep, MatchTier, ScorerId } from "./types";
import { wratioScore } from "./rapidfuzz";

const STOP_TOKENS = new Set([
  "with", "and", "the", "a", "an", "of", "for", "in", "on", "w", "ea", "pkg", "pack",
  "value", "fresh", "food", "foods", "mart", "label", "superior", "bridge",
]);

const GENERIC_PRODUCT_TOKENS = new Set([
  "rice", "ball", "balls", "pork", "soy", "sauce", "milk", "egg", "eggs", "tofu",
  "chips", "tomato", "tomatoes", "pasta", "frozen", "fresh", "boneless", "whole",
  "thyme", "herb", "herbs", "spice", "spices",
]);

const EMB_RAPIDFUZZ_EMB_WEIGHT = 0.6;
const EMB_RAPIDFUZZ_RF_WEIGHT = 0.4;
const PRODUCTION_EMB_WEIGHT = 0.75;
const PRODUCTION_TOK_WEIGHT = 0.25;

export interface ScorePairOptions {
  excludeSamePhoto?: boolean;
  scorerMode?: ScorerId;
}

export function normalizeName(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9\u4e00-\u9fff]+/g, " ")
    .trim()
    .split(/\s+/)
    .filter(Boolean)
    .join(" ");
}

function tokenize(name: string): string[] {
  return normalizeName(name)
    .split(" ")
    .filter((t) => t && !STOP_TOKENS.has(t));
}

function tokenJaccard(a: string, b: string, ignoreGeneric = false): number {
  const ta = new Set(
    tokenize(a).filter((t) => !(ignoreGeneric && GENERIC_PRODUCT_TOKENS.has(t))),
  );
  const tb = new Set(
    tokenize(b).filter((t) => !(ignoreGeneric && GENERIC_PRODUCT_TOKENS.has(t))),
  );
  if (ta.size === 0 || tb.size === 0) return 0;
  let intersection = 0;
  for (const t of ta) {
    if (tb.has(t)) intersection += 1;
  }
  const union = ta.size + tb.size - intersection;
  return union === 0 ? 0 : intersection / union;
}

function cosineSimilarity(a: number[], b: number[]): number {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i += 1) {
    dot += a[i]! * b[i]!;
    normA += a[i]! * a[i]!;
    normB += b[i]! * b[i]!;
  }
  if (normA === 0 || normB === 0) return 0;
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

function embeddingScore(vectorA: number[] | null | undefined, vectorB: number[] | null | undefined): number {
  if (!vectorA || !vectorB) return 0;
  return (cosineSimilarity(vectorA, vectorB) + 1) / 2;
}

function round(n: number, digits = 4): number {
  const factor = 10 ** digits;
  return Math.round(n * factor) / factor;
}

function finalizeResult(
  source: MatchSighting,
  target: MatchSighting,
  final: number,
  tier: MatchTier,
  steps: MatchStep[],
  minScore: number,
  samePhoto: boolean,
  excludeSamePhoto: boolean,
): MatchDetail {
  const skipped = samePhoto && excludeSamePhoto;
  return {
    source_id: source.id,
    target_id: target.id,
    final_score: final,
    tier,
    skipped,
    skip_reason: skipped ? "Same photo (excluded in production related search)" : undefined,
    above_threshold: !skipped && final >= minScore,
    steps,
  };
}

export function scorePairDetail(
  source: MatchSighting,
  target: MatchSighting,
  vectorA: number[] | null | undefined,
  vectorB: number[] | null | undefined,
  minScore: number,
  options: ScorePairOptions = {},
): MatchDetail {
  const excludeSamePhoto = options.excludeSamePhoto ?? true;
  const scorerMode = options.scorerMode ?? "production";
  const steps: MatchStep[] = [];
  const samePhoto = source.photo_id === target.photo_id;

  if (source.id === target.id) {
    return {
      source_id: source.id,
      target_id: target.id,
      final_score: 0,
      tier: "skipped",
      skipped: true,
      skip_reason: "Same product",
      above_threshold: false,
      steps: [{ id: "skip", label: "Skipped", value: "self" }],
    };
  }

  if (samePhoto) {
    steps.push({
      id: "same_photo",
      label: "Same photo",
      value: true,
      note: excludeSamePhoto
        ? "Production matcher excludes co-photo products from related search"
        : "Included for full pairwise matrix",
    });
  }

  steps.push(
    { id: "source_name", label: "Source name", value: source.product_name },
    { id: "target_name", label: "Target name", value: target.product_name },
    { id: "source_barcode", label: "Source barcode", value: source.barcode ?? null },
    { id: "target_barcode", label: "Target barcode", value: target.barcode ?? null },
  );

  if (source.barcode && target.barcode && source.barcode === target.barcode) {
    steps.push({ id: "path", label: "Match path", value: "barcode" });
    return finalizeResult(source, target, 1.0, "barcode", steps, minScore, samePhoto, excludeSamePhoto);
  }

  const normA = normalizeName(source.product_name);
  const normB = normalizeName(target.product_name);
  steps.push(
    { id: "normalized_source", label: "Normalized source", value: normA },
    { id: "normalized_target", label: "Normalized target", value: normB },
  );

  if (normA === normB) {
    steps.push({ id: "path", label: "Match path", value: "exact_name" });
    return finalizeResult(source, target, 1.0, "exact_name", steps, minScore, samePhoto, excludeSamePhoto);
  }

  const compositeTier: MatchTier =
    scorerMode === "emb_rapidfuzz" ? "composite_emb_rapidfuzz" : "composite_embedding";
  steps.push({ id: "scorer_mode", label: "Scorer", value: scorerMode });
  steps.push({ id: "path", label: "Match path", value: compositeTier });

  const rawCosine = vectorA && vectorB ? cosineSimilarity(vectorA, vectorB) : null;
  const emb = embeddingScore(vectorA, vectorB);

  let composite: number;
  if (scorerMode === "emb_rapidfuzz") {
    const rapidfuzz = wratioScore(source.product_name, target.product_name);
    composite = EMB_RAPIDFUZZ_EMB_WEIGHT * emb + EMB_RAPIDFUZZ_RF_WEIGHT * rapidfuzz;
    steps.push(
      { id: "embedding_cosine", label: "Embedding cosine similarity", value: rawCosine === null ? null : round(rawCosine) },
      { id: "embedding_score", label: "Embedding score (cosine+1)/2", value: round(emb) },
      { id: "rapidfuzz_wratio", label: "RapidFuzz WRatio", value: round(rapidfuzz) },
      {
        id: "composite_formula",
        label: "Composite formula",
        value: `${EMB_RAPIDFUZZ_EMB_WEIGHT} * emb + ${EMB_RAPIDFUZZ_RF_WEIGHT} * rapidfuzz`,
      },
      { id: "composite_raw", label: "Composite raw", value: round(composite) },
    );
  } else {
    const tok = tokenJaccard(source.product_name, target.product_name, true);
    composite = PRODUCTION_EMB_WEIGHT * emb + PRODUCTION_TOK_WEIGHT * tok;
    steps.push(
      { id: "embedding_cosine", label: "Embedding cosine similarity", value: rawCosine === null ? null : round(rawCosine) },
      { id: "embedding_score", label: "Embedding score (cosine+1)/2", value: round(emb) },
      { id: "token_jaccard", label: "Token Jaccard (ignore generic)", value: round(tok) },
      {
        id: "composite_formula",
        label: "Composite formula",
        value: `${PRODUCTION_EMB_WEIGHT} * emb + ${PRODUCTION_TOK_WEIGHT} * tok`,
      },
      { id: "composite_raw", label: "Composite raw", value: round(composite) },
    );
  }

  steps.push({ id: "final_score", label: "Final score", value: round(composite) });
  return finalizeResult(
    source,
    target,
    composite,
    compositeTier,
    steps,
    minScore,
    samePhoto,
    excludeSamePhoto,
  );
}

export function toSighting(product: {
  id: string;
  photo_id: string;
  product_name: string;
  category: string;
  brand?: string | null;
  barcode?: string | null;
}): MatchSighting {
  return {
    id: product.id,
    photo_id: product.photo_id,
    product_name: product.product_name,
    category: product.category,
    brand: product.brand ?? null,
    barcode: product.barcode ?? null,
  };
}

export function formatScore(score: number): string {
  return score.toFixed(3);
}

export function tierLabel(tier: MatchTier): string {
  switch (tier) {
    case "barcode":
      return "Barcode";
    case "exact_name":
      return "Exact name";
    case "composite_embedding":
      return "Composite + embedding";
    case "composite_emb_rapidfuzz":
      return "Emb 0.6 + RapidFuzz 0.4";
    case "skipped":
      return "Skipped";
    default:
      return tier;
  }
}
